geo-traits/src/coord.rs

```diff
@@ -0,0 +1,141 @@
+use std::marker::PhantomData;
+
+use geo_types::{Coord, CoordNum};
+
+use crate::Dimensions;
+
+/// A trait for accessing data from a generic Coord.
+///
+/// Refer to [geo_types::Coord] for information about semantics and validity.
+pub trait CoordTrait {
+    /// The coordinate type of this geometry
+    type T: CoordNum;
+
+    /// Dimensions of the coordinate tuple
+    fn dim(&self) -> Dimensions;
+
+    /// Access the n'th (0-based) element of the CoordinateTuple.
+    /// Returns NaN if `n >= DIMENSION`.
+    /// See also [`nth_unchecked()`](Self::nth_unchecked).
+    fn nth(&self, n: usize) -> Option<Self::T> {
+        if n < self.dim().size() {
+            Some(self.nth_unchecked(n))
+        } else {
+            None
+        }
+    }
+
+    /// x component of this coord.
+    fn x(&self) -> Self::T;
+
+    /// y component of this coord.
+    fn y(&self) -> Self::T;
+
+    /// Returns a tuple that contains the x/horizontal & y/vertical component of the coord.
+    fn x_y(&self) -> (Self::T, Self::T) {
+        (self.x(), self.y())
+    }
+
+    /// Access the n'th (0-based) element of the CoordinateTuple.
+    /// May panic if n >= DIMENSION.
+    /// See also [`nth()`](Self::nth).
+    fn nth_unchecked(&self, n: usize) -> Self::T;
```

Comment:
When I first read this, I thought this would behave similarly to `Vec#get_unchecked` which is an `unsafe` method which does not do a bounds check, hence it being marked as `unsafe`. In fact, it looks like all functions named `*_unchecked` Rust's standard library are marked as `unsafe`, which to me implies they skip the bounds check.

My first though was, what if we had a naming convention like this:

```rust
fn nth_checked(n) -> Option<T> {}
fn nth(n) -> T {} // Panics
unsafe fn nth_unchecked(n) -> T
```

But that may psychologically push people to use the panic'ing version.

I don't know what the right answer is, just wanted to call out that I was surprised when implementing `Coord` that `nth_unchecked` was not marked with `unsafe`.

Reply:
Another option 🤷🏻 

```rust
fn nth(n) -> Option<T> {}
fn nth_or_panic(n) -> T {} // Panics
unsafe fn nth_unchecked(n) -> T
```

Reply:
But we also definitely don't need to have an `unsafe` method, in fact probably better to leave it out for this initial implementation

Reply:
```
fn nth(n) -> Option<T> {}
fn nth_or_panic(n) -> T {} // Panics
```
That's nice. 👏 

```diff
- coord.nth_unchecked(3);
+ coord.nth(3).unwrap()
```

We could also just leave out the `panic` flavor.

Reply:
Although I didn't find it in the stdlib, apparently there's quite a bit of precedent for `_or_panic` in 3rd party crates
 https://github.com/search?q=or_panic+language%3ARust&type=code

Reply:
Open to suggestions here. I suppose it's worthwhile having a variant that is actually unsafe (e.g. doesn't even do the bounds check), for situations where you know you're in bounds (like the iterator)